File: src/b2500_meter/powermeter/emlog.py
import aiohttp

from .base import Powermeter


class Emlog(Powermeter):
    def __init__(self, ip: str, meterindex: str, json_power_calculate: bool):
        self.ip = ip
        self.meterindex = meterindex
        self.json_power_calculate = json_power_calculate
        self.session: aiohttp.ClientSession | None = None

    async def start(self) -> None:
        if self.session:
            return
        self.session = aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=10))

    async def stop(self) -> None:
        if self.session:
            await self.session.close()
            self.session = None

    async def get_json(self, path):
        if not self.session:
            raise RuntimeError("Session not started; call start() first")
        url = f"http://{self.ip}{path}"
        async with self.session.get(url) as resp:
            return await resp.json(content_type=None)

    async def get_powermeter_watts_async(self) -> list[float]:
        response = await self.get_json(
            f"/pages/getinformation.php?heute&meterindex={self.meterindex}"
        )
        if not self.json_power_calculate:
            return [int(response["Leistung170"])]
        else:
            power_in = response["Leistung170"]
            power_out = response["Leistung270"]
            return [int(power_in) - int(power_out)]

File: crates/powermeters/src/emlog.rs
//! `EMLOG` — port of `src/astrameter/powermeter/emlog.py`.

use std::sync::Arc;
use std::time::Duration;

use astrameter_config::Section;
use astrameter_core::{Error, Powermeter, Result};
use astrameter_platform::{
    http::{HttpClient, HttpRequest},
    Platform,
};
use async_trait::async_trait;
use serde_json::Value;

pub struct Emlog {
    ip: String,
    meter_index: String,
    power_calculate: bool,
    http: Arc<dyn HttpClient>,
}

fn coerce_int(v: &Value) -> Result<i64> {
    match v {
        Value::Number(n) => n
            .as_i64()
            .or_else(|| n.as_f64().map(|f| f as i64))
            .ok_or_else(|| Error::decode("not an int")),
        Value::String(s) => s
            .trim()
            .parse::<i64>()
            .or_else(|_| s.trim().parse::<f64>().map(|f| f as i64))
            .map_err(|e| Error::decode(format!("not an int: {e}"))),
        _ => Err(Error::decode(format!("can't coerce {v} to int"))),
    }
}

#[async_trait]
impl Powermeter for Emlog {
    async fn get_powermeter_watts(&self) -> Result<Vec<f64>> {
        let url = format!(
            "http://{}/pages/getinformation.php?heute&meterindex={}",
            self.ip, self.meter_index
        );
        let resp = self
            .http
            .request(HttpRequest::get(&url).with_timeout(Duration::from_secs(10)))
            .await
            .map_err(|e| Error::transport(format!("emlog: {e}")))?;
        let json: Value = serde_json::from_slice(&resp.body)
            .map_err(|e| Error::decode(format!("emlog json: {e}")))?;
        let power_in = coerce_int(
            json.get("Leistung170")
                .ok_or_else(|| Error::decode("emlog: missing Leistung170"))?,
        )?;
        if !self.power_calculate {
            return Ok(vec![power_in as f64]);
        }
        let power_out = coerce_int(
            json.get("Leistung270")
                .ok_or_else(|| Error::decode("emlog: missing Leistung270"))?,
        )?;
        Ok(vec![(power_in - power_out) as f64])
    }
}

pub fn create(section: &Section<'_>, platform: Arc<Platform>) -> Result<Arc<dyn Powermeter>> {
    Ok(Arc::new(Emlog {
        ip: section.get_required("IP")?.to_string(),
        meter_index: section.get_string("METER_INDEX", ""),
        power_calculate: section.get_bool("JSON_POWER_CALCULATE", false)?,
        http: platform.http.clone(),
    }))
}
